### ISSUE TITLE
Error trying to use 2.6.149  (Error extracting file)

### ISSUE THREAD

**Issue by beelee1** [triplea-game/triplea#9252]
Error extracting file: /tmp/triplea6560939776671296847/Global_40_House_Rules.zip, /tmp/triplea-unzip2861211619708353438/global_40_house_rules-master, {}  java.nio.file.DirectoryNotEmptyException: /tmp/triplea-unzip2861211619708353438/global_40_house_rules-master  If this problem happens frequently and is something you cannot fix, please report it to TripleA 

Trying to replicate this error #9250 but i can't dl map. I get a bunch of errors on startup trying to unzip all the maps in dl maps. Which maxed out my "root", which i had to spend a couple hours figuring out what and how to purge so it'd work again.

So basically, for me, This is unusable

Edit
I removed the map and tried to dl it and it'd dl but not showup. I then manually unzipped the map and put the folder directly in my dl maps and that didn't work either.

**Comment by DanVanAtta** [comment]
Do you wind up with a folder named something like "Global_40_House_Rules" in your downloaded maps folder?

**Comment by beelee1** [comment]
So, I tried this again and no folder showed up. I removed the zip and no folder showed up. When I choose it in DL maps, it throws the above error. It will then not show up as a DL option even though it's not in my downloaded maps folder.

I close DL maps and then it'll show up as a DL option again. I have a map I'm working on that is in a folder, no zip and it shows up and works correctly.

I just tried unzipping and loading the map manually and it shows up now. Idk y it didn't earlier. I must've screwed something up when I first tried it.

**Comment by DanVanAtta** [comment]
Extraction from zip to folder occurs at startup.
If there is no zip, there is nothing to extract to a folder.

DL listing is based on the folders that are present and specifically what is listed in the 'map.yml' files in those folders.  There *could* be an issue here where at present maps do not have a 'map.yml' file, that file is generated on startup, so you could run into an inconsistent state after downloading a map and it then not having a map.yml file. If you re-open the downloader, it might not think you have the map downloaded anymore as it can't find the corresponding map.yml file (even though you have the have the map extracted), at which point you run into the above error as it'll download a zip and then try to incorrectly overwrite the previous.

To help this from a game engine perspective, we can do the following:
1. We need to finally get map.yml files uploaded to all maps
2. The engine should be checking if the map folder already exists and give a better error message or just go ahead and overwrite it. Probably overwrite it as we want the downloader to update maps and not just install maps.

**Comment by beelee1** [comment]
right on. Just to clarify, when I first fired up 2.6.149, I had all my existing maps from 2.5 in "downloadedMaps". The biggest issue was "root" getting maxed out. Took me a while to figure that out, so trying to save others from the same experience.

**Comment by DanVanAtta** [comment]
That would be worthwhile for us to investigate as well, notably if we can ensure that folders are unzipped one at a time with the zip removed after each one and second that we can better fail if there is insufficient disk space.

**Comment by DanVanAtta** [comment]
This could potentially be fixed now in the very latest. The map update logic was not deleting existing maps before updating. 
@beelee1 would you mind checking if you are still seeing this problem?

**Comment by beelee1** [comment]
@DanVanAtta I'll try and check out in a bit.

**Comment by emmanuel-perlow** [comment]
Just checked it, the behaviour reported by @sumpfralle still happens on 2.6.379 (https://github.com/triplea-game/triplea/commit/a11fe86268fab7c7bf4ce4b32e44890aca1ac7bd) on Linux.

In case a workaround is needed:
You can manually move the map folder in `/tmp/triplea-unzipXXXXX/` to `${HOME}/triplea/downloadedMaps/` (You may need to close the "Select Game"/"Download Maps" dialog and reopen it, to see the new map).

**Comment by beelee1** [comment]
just tried this in 2.6.574 pre and it still happens

**Comment by DanVanAtta** [comment]
What are the exact sequence of actions you take after launching the game to then observe this error @beelee1 ? Are you simply launching the game?

What is the exact text message of the error?

Could you please check the log file generated by triple for a stack trace and attach that log file to this issue?

**Comment by beelee1** [comment]
@DanVanAtta  yea each individual map shows up as separate errors. I'll fire it up again and post a few. back in a few minutes

**Comment by beelee1** [comment]
@DanVanAtta I'll clear some space and try this again. Maybe it only does it the first time with the consecutive map errors and if you close it and restart then it does this. I'll try again in a bit

**Comment by DanVanAtta** [comment]
@beelee1 , the attached log file is empty. Recommend you just move all your maps into some other folder & delete the contents of /tmp, then try again.

**Comment by RoiEXLab** [comment]
@beelee1 Looks like this issue has been taken care of, closing this...

**Comment by asvitkine** [comment]
@beelee1 mentioned that this is still not working for them on: https://forums.triplea-game.org/topic/2998/2-6-testing-report-2-6-problems-here/54?page=3

@beelee1 - seems you tried to post some screenshots on Feb 17, but it just looks like some file paths and not the actual images. Do you mind trying to post them again?

If this is preventing you from trying/using 2.6, I'd like us to figure out the problem and fix it.

**Comment by beelee1** [comment]
Here's another attempt at the log
[triplea log.zip](https://github.com/triplea-game/triplea/files/8782174/triplea.log.zip)
 looks like it worked this time. I'll post the error pics later

**Comment by asvitkine** [comment]
Thanks!

So looks like we're getting an error from `ensureDirectoryEmpty()` in Unix FS code, which does:
```
        if (sourceAttrs.isDirectory()) {
            ensureEmptyDir(source);
            copyDirectory(source, sourceAttrs, target, flags);
```

So seems it requires the source directory to be empty before moving? That seems... surprising to me.

The stack trace:

```
java.nio.file.DirectoryNotEmptyException: /tmp/triplea-unzip3447058957441722647/global_40_expansion-master
	at java.base/sun.nio.fs.UnixCopyFile.ensureEmptyDir(UnixCopyFile.java:384)
	at java.base/sun.nio.fs.UnixCopyFile.move(UnixCopyFile.java:484)
	at java.base/sun.nio.fs.UnixFileSystemProvider.move(UnixFileSystemProvider.java:267)
	at java.base/java.nio.file.Files.move(Files.java:1422)
	at org.triplea.io.FileUtils$FileMoveOperation.move(FileUtils.java:265)
	at org.triplea.io.FileUtils.replaceFolder(FileUtils.java:276)
	at org.triplea.io.FileUtils.replaceFolder(FileUtils.java:259)
	at games.strategy.engine.framework.map.file.system.loader.ZippedMapsExtractor.unzipMapThrowing(ZippedMapsExtractor.java:144)
	at games.strategy.engine.framework.map.file.system.loader.ZippedMapsExtractor.unzipMap(ZippedMapsExtractor.java:111)
	at games.strategy.engine.framework.map.file.system.loader.ZippedMapsExtractor.lambda$unzipMapFiles$2(ZippedMapsExtractor.java:55)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
```

**Comment by asvitkine** [comment]
Documentation for `Files.move()` which we're using says:

```This method may be invoked to move an empty directory. In some implementations a directory has entries for special files or links that are created when the directory is created. In such implementations a directory is considered empty when only the special entries exist. When invoked to move a directory that is not empty then the directory is moved if it does not require moving the entries in the directory. For example, renaming a directory on the same [FileStore](https://docs.oracle.com/javase/7/docs/api/java/nio/file/FileStore.html) will usually not require moving the entries in the directory. When moving a directory requires that its entries be moved then this method fails (by throwing an IOException). To move a file tree may involve [copy](https://docs.oracle.com/javase/7/docs/api/java/nio/file/Files.html#copy(java.nio.file.Path,%20java.nio.file.Path,%20java.nio.file.CopyOption...))ing rather than moving directories and this can be done using the copy method in conjunction with the [Files.walkFileTree](https://docs.oracle.com/javase/7/docs/api/java/nio/file/Files.html#walkFileTree(java.nio.file.Path,%20java.util.Set,%20int,%20java.nio.file.FileVisitor)) utility method.
```

**Comment by asvitkine** [comment]
@beelee1 Do you know if your downloadedMaps and /tmp directories are on different partitions?

That could explain the error.
Perhaps we can make the temp dir in a subdir of downloadedMaps.

**Comment by asvitkine** [comment]
I sent https://github.com/triplea-game/triplea/pull/10512 - I don't have a way to test, but I suspect it should resolve this problem. Please confirm once that PR is merged into a prerelease!

**Comment by beelee1** [comment]
@asvitkine right on. I'll give it a shot Thanks

**Comment by beelee1** [comment]
@asvitkine Sweet ! Works like a charm : ) Knew it had been a while but didn't realize it was a year. Be nice to use the prerelease again :)

One thing I noticed, is it unzips everything now, so the first time you run it, it'll take a minute if you have a lot of maps. From then on, it'll fire right up. Some people might not be aware of that.

Once again BIG Thank You !!!

**Comment by beelee1** [comment]
Yea so the game behavior works correctly with  #10749 fix with 2.6.838 but the updated notes don't show up. The fix was in the units folder. The xml changes don't show up though. Which were Game Version/Notes only.

So Idk if that means 2.6 can't see any xml changes or not. I'm gonna do some xml updates to G 40 Expansion and then try with 2.5 and latest stable to see if they work.

Here's a save
[Harbor 12 works but Notes don't.zip](https://github.com/triplea-game/triplea/files/8983968/Harbor.12.works.but.Notes.don.t.zip)
Should be 2.72 in the Notes

@asvitkine gonna reopen

Also, I tested version 2.71 and confirmed the error with issue #10729

**Comment by beelee1** [comment]
yea so the Notes are all jacked up on 2.6.838 Here's 2.5

![Screenshot from 2022-06-25 02-15-33](https://user-images.githubusercontent.com/24382682/175763590-eef24976-7d82-4afb-be79-6b817a17e69f.png)

![Screenshot from 2022-06-25 02-16-28](https://user-images.githubusercontent.com/24382682/175763625-22970e08-e003-4bb3-9150-0b650fdcc05c.png)

Here's 2.6838

![Screenshot from 2022-06-25 02-17-03](https://user-images.githubusercontent.com/24382682/175763707-09f67c8f-e5cb-4b5d-a21b-7dee1148f961.png)

![Screenshot from 2022-06-25 02-19-09](https://user-images.githubusercontent.com/24382682/175763666-feb30c5c-15fc-4432-b3b8-0ecbef760e2c.png)

Notifications worked correctly and the xml change other than the Notes worked, but I'm going to run another test, an obvious one : ) , for that to make sure.

I wonder if the Notes change has anything to do with it. Not only do you have to scroll the other way and make an additional click, it blocks your other windows. You can't click on anything, you have to close Notes first. Quite annoying.

At any rate, I'll test some more.

**Comment by beelee1** [comment]
I changed Infantry from "Attack 1" to "Attack 5" and it worked correctly on 2.5 as well as 2.6.838 so "Game Notes" and "Map Version" are what are not showing up correctly.

The map version doesn't show before game start in the pre as it does in the stable either.
Stable
![Screenshot from 2022-06-25 03-06-28](https://user-images.githubusercontent.com/24382682/175764728-963febb6-468a-4ce8-b4f8-0c850a7cb9c4.png)
Pre
![Screenshot from 2022-06-25 03-05-19](https://user-images.githubusercontent.com/24382682/175764758-44e6cc02-7851-450d-a08f-7c063298f16f.png)

Idk if any of that is related but ... Maybe the fact that the new downloaded map isn't starting as azip or at least it doesn't seem too, has something to do with it ?

Anyway, let me know if you want me to test something else.

**Comment by beelee1** [comment]
I'm gonna close this and open a new one on the "Notes". I still think the two are related but w/e will close for now
